### PR TITLE
Initial simple openmetricstest tool implementation

### PR DESCRIPTION
**Pull request by robskillington**
Really basic but an easy way to test whether parsing fails/passes for test cases in a testdata directory.

Here's an example testing `true` and `false`:
```sh
# Compile from repository root
make openmetricstest

# Run and since echo will always accept any stdin observe success
./bin/openmetricstest -testdata-dir ./tests/testdata/parsers/simple_counter -cmd-parser-text true
2019/04/09 21:25:56 RUN test: simple_counter
2019/04/09 21:25:56 parse-result-validator ok
2019/04/09 21:25:56 PASS test: simple_counter
2019/04/09 21:25:56 OK passed=1, failed=0, total_failures=0

# Run and since the script will always fail regardless of stdin observe failure
./bin/openmetricstest -testdata-dir ./tests/testdata/parsers/simple_counter -cmd-parser-text false
2019/04/09 21:40:30 RUN test: simple_counter
2019/04/09 21:40:30 parse-result-validator error:
2019/04/09 21:40:30 > parse error: exit status 1
2019/04/09 21:40:30 FAIL test: simple_counter
2019/04/09 21:40:30 FAILED passed=0, failed=1, total_failures=1
```

### CODE REVIEW

File: tests/openmetricstest/main.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+	"path"
+	"strings"
+)
+
+const (
+	testFileName = "test.json"
+)
+
+var (
+	testDataDirArg       = flag.String("testdata_dir", "testdata", "testdata directory to use")
```
Comment (brian-brazil) [inline tests/openmetricstest/main.go]: Hyphens and periods are what we usually use

File: tests/openmetricstest/main.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	cmdTestParserTextArg = flag.String("cmd_test_parser_text", "", "command to run to test parser in text mode")
+)
+
+func main() {
+	flag.Parse()
+
+	testDataDir := strings.TrimSpace(*testDataDirArg)
```
Comment (brian-brazil) [inline tests/openmetricstest/main.go]: We don't usually do this, if the user has a space we should use it

File: tests/openmetricstest/main.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		log.Fatalf("cannot read testdata dir: %v", err)
+	}
+
+	for _, file := range files {
+		if file.Name() == testFileName {
+			runTest(dir, opts)
+			continue
+		}
+		if file.IsDir() {
+			// Recurse
```
Comment (brian-brazil) [inline tests/openmetricstest/main.go]: You would use `filepath.Walk` for this

File: tests/openmetricstest/main.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+			runTests(path.Join(dir, file.Name()), opts)
+		}
+	}
+}
+
+type testDef struct {
+	Parse   []parseDef   `json:"parse"`
```
Comment (brian-brazil) [inline tests/openmetricstest/main.go]: This seems overly nested, I'd put fold it into one struct

File: tests/openmetricstest/main.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+			if err := cmd.Start(); err != nil {
+				log.Fatalf("cannot start cmd: %v", err)
+			}
+
+			if _, err := stdin.Write(input); err != nil {
+				log.Fatalf("cannot write input: %v", err)
+			}
+
+			if err := validateTestOutcome(test, cmd); err != nil {
+				log.Fatalf("failed test: test=%s, err=%v", dir, err)
```
Comment (brian-brazil) [inline tests/openmetricstest/main.go]: Might it be better to collect all the failures, so you don't have to get them one at a time?

File: tests/testdata/parsers/simple_counter/test.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{
+  "parse": [
+    {
+      "type": "text",
+      "file": "metrics.txt"
```
Comment (brian-brazil) [inline tests/testdata/parsers/simple_counter/test.json]: I was thinking more that we'd inline the metrics here

File: tests/openmetricstest/main.go (hunk 1 above)
Comment (robskillington) [reply tests/openmetricstest/main.go]: Sounds good, will use that.

File: tests/openmetricstest/main.go (hunk 2 above)
Comment (robskillington) [reply tests/openmetricstest/main.go]: Sounds good, will remove the trim.

File: tests/openmetricstest/main.go (hunk 3 above)
Comment (robskillington) [reply tests/openmetricstest/main.go]: Good call, will use idiomatic filepath.Walk instead.

File: tests/openmetricstest/main.go (hunk 4 above)
Comment (robskillington) [reply tests/openmetricstest/main.go]: Fair, we can always add more later if we want multi-input and multi-result-checking support.

File: tests/openmetricstest/main.go (hunk 5 above)
Comment (robskillington) [reply tests/openmetricstest/main.go]: Agreed, will update.

File: tests/testdata/parsers/simple_counter/test.json (hunk 6 above)
Comment (robskillington) [reply tests/testdata/parsers/simple_counter/test.json]: Let's discuss this tomorrow, the issue becomes that for proto file contents we won't be able to put binary data inline.  

Also for testing edge cases with weird cases, it's probably better not to go through a JSON parse step which could do some strange translation before running the test case I would think.

**Comment by robskillington** [comment]
Address feedback, one comment still to discuss was my response to the request to inline the text for the test, here was my response:
> Perhaps we should discuss this during the biweekly call, the issue is that for testing edge cases with strange characters, it's probably better not to go through a JSON parse step which could do some strange translation before running the test case I would think. 

> Also for testing proto contents we won't be able to put binary data inline.

**Comment by robskillington** [comment]
@brian-brazil this is good for another review.

I was wondering what CI we should add to this repository, CNCF has some CI support I had heard - should I investigate that?  (Just for purely making sure these tools builds and passes its own tests, etc).

**Comment by brian-brazil** [comment]
CircleCI would be best I think.

**Comment by robskillington** [comment]
Okay great, @brian-brazil I added a makefile and Circle CI config - I don't have access to add the repository to Circle CI though.

If someone could go through adding the repository to Circle CI then it technically should be able to run on this PR once it's ready.

File: src/cmd/openmetricstest/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+2019/04/09 21:25:56 RUN test: simple_counter
+2019/04/09 21:25:56 parse-result-validator ok
+2019/04/09 21:25:56 PASS test: simple_counter
+2019/04/09 21:25:56 OK passed=1, failed=0, total_failures=0
+```
+
+### Testing text parsing with a script that always fails
+
+```
+# Create a shell script that will always fail on any stdin
```
Comment (brian-brazil) [inline src/cmd/openmetricstest/README.md]: `false` would be simpler.

File: src/cmd/openmetricstest/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+```
+Usage of ./openmetricstest:
+  -cmd-parser-text string
+        command to run to test parser in text mode
+  -testdata-dir string
+        testdata directory to use (default "./tests/testdata")
+```
+
+## Testing an implementation
```
Comment (brian-brazil) [inline src/cmd/openmetricstest/README.md]: It could be clearer what the contract is, that we'll pass the input to a binary and it should succeed/fail accordingly.

File: src/cmd/openmetricstest/main.go [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+	for _, elem := range r.tests {
+		if len(elem.failures) == 0 {
+			success++
+		} else {
+			failed++
+		}
+		failures += len(elem.failures)
+	}
+
+	result := fmt.Sprintf("passed=%d, failed=%d, total_failures=%d\n",
```
Comment (brian-brazil) [inline src/cmd/openmetricstest/main.go]: This looks like a double `\n`

File: src/cmd/openmetricstest/main.go [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+		testDir, _ := path.Split(filePath)
+		result, err := runTest(testDir, opts)
+		results.tests = append(results.tests, result)
+		return err
+	})
+	return results, err
+}
+
+type testDef struct {
```
Comment (brian-brazil) [inline src/cmd/openmetricstest/main.go]: This can be flattened more.

File: src/cmd/openmetricstest/README.md (hunk 7 above)
Comment (robskillington) [reply src/cmd/openmetricstest/README.md]: Good point, I'll update to use `false` instead.

File: src/cmd/openmetricstest/README.md (hunk 8 above)
Comment (robskillington) [reply src/cmd/openmetricstest/README.md]: Good point, I'll update that.

File: src/cmd/openmetricstest/main.go (hunk 9 above)
Comment (robskillington) [reply src/cmd/openmetricstest/main.go]: Good catch, thanks.

**Comment by robskillington** [comment]
@SuperQ if you don't mind weighing in on the basic Circle CI config `.circleci/config.yml` and the use of `Makefile` that'd be greatly appreciated!

File: src/cmd/openmetricstest/main.go (hunk 10 above)
Comment (robskillington) [reply src/cmd/openmetricstest/main.go]: Sounds good.

File: src/cmd/openmetricstest/main.go [hunk 11]
```diff
@@ -102,26 +103,18 @@ func runTests(dir string, opts runTestsOptions) (runTestsResults, error) {
...
-
-type outcomeDef struct {
-	ParseResult *parseResultDef `json:"parseResult"`
+	Type        string          `json:"type"`
+	File        string          `json:"file"`
+	ResultParse *resultParseDef `json:"resultParse"`
 }
 
-type parseResultDef struct {
+type resultParseDef struct {
```
Comment (brian-brazil) [inline src/cmd/openmetricstest/main.go]: Why not put this directly in the outcomeDef?

File: src/cmd/openmetricstest/main.go [hunk 12]
```diff
@@ -102,26 +103,18 @@ func runTests(dir string, opts runTestsOptions) (runTestsResults, error) {
...
 	Valid bool `json:"valid"`
```
Comment (brian-brazil) [inline src/cmd/openmetricstest/main.go]: shouldParse would be clearer

File: Makefile [hunk 13]
```diff
@@ -0,0 +1,46 @@
+BUILD := $(abspath ./bin)
+BINARIES :=         \
+	openmetricstest \
+
+# To test "echo": 
+# make test-impl cmd-parser-text=echo
+# To test failure of "false":
```
Comment (brian-brazil) [inline Makefile]: I'm unclear why there's a fail parameter

Comment (robskillington) [reply Makefile]: Hey, this is so we can ensure that implementations we expect to fail do fail. (i.e. inverse the result).

This was mainly for CI, although I could try and use `bash -c "! make test-impl cmd-parser-text=false"` and see if that returns success on failure instead?

Comment (robskillington) [reply Makefile]: I'll remove it and once CircleCI is working see if we can't just do something like I suggested.

File: src/cmd/openmetricstest/main.go (hunk 12 above)
Comment (robskillington) [reply src/cmd/openmetricstest/main.go]: Sure thing.

File: src/cmd/openmetricstest/main.go (hunk 11 above)
Comment (robskillington) [reply src/cmd/openmetricstest/main.go]: Sure, it's just in the future I wanted to support adding multiple types of verification (parsing just being one kind, exposition/serialization being another).

I'll just remove it if we don't want support for multiple types of result verifiers for now.

**Comment by robskillington** [comment]
@brian-brazil this is ready for another round, ty.

File: src/cmd/openmetricstest/main.go [hunk 14]
```diff
@@ -103,18 +103,9 @@ func runTests(dir string, opts runTestsOptions) (runTestsResults, error) {
...
-type resultParseDef struct {
-	Valid bool `json:"valid"`
-}
-
-// validator returns the concrete validator for this test result definition
-func (d resultParseDef) validator() testResultValidator {
-	return resultParseValidator{def: d}
+	Type        string `json:"type"`
+	File        string `json:"file"`
+	ShouldParse *bool  `json:"shouldParse"`
```
Comment (brian-brazil) [inline src/cmd/openmetricstest/main.go]: Why the * ?

File: src/cmd/openmetricstest/README.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+This command specifies the process to run to test parsing OpenMetrics text exposition test cases.  It should take the OpenMetrics text exposition as stdin.
+
+## Examples
+
+These examples are from running the tool in the repository root after building it.
+
+### Testing text parsing with "echo"
+
+This is an example of tersting that "echo" can parse a simple counter text exposition:
```
Comment (brian-brazil) [inline src/cmd/openmetricstest/README.md]: testing

true might be a better choice

File: src/cmd/openmetricstest/main.go (hunk 14 above)
Comment (robskillington) [reply src/cmd/openmetricstest/main.go]: Mainly so I know whether to run the parse validator or not, ideally in the future we have other validators and whether I should run them depends on whether the test definition has set the option or not.

File: src/cmd/openmetricstest/README.md (hunk 15 above)
Comment (robskillington) [reply src/cmd/openmetricstest/README.md]: Will update to `true` and thanks for the spelling catch.

**Comment by robskillington** [comment]
@brian-brazil ready for another round.

File: src/cmd/openmetricstest/main.go (hunk 14 above)
Comment (brian-brazil) [reply src/cmd/openmetricstest/main.go]: That seems a bit overengineered for what we're currently doing, tristates can be confusing. I'd say keep it simple for now.

**Comment by robskillington** [comment]
@brian-brazil addressed pointer comment.
